enterprise/internal/campaigns/syncer.go

```diff
@@ -307,80 +372,124 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
-	if q.cancel != nil {
-		// cancel closes the done chan on the existing context
-		// which will cause the existing worker routine to exit
-		q.cancel()
+func (pq *changesetPriorityQueue) Swap(i, j int) {
+	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
+	pq.index[pq.items[i].changesetID] = i
+	pq.index[pq.items[j].changesetID] = j
+}
+
+// Push is here to implement the Heap interface, please use Upsert
+func (pq *changesetPriorityQueue) Push(x interface{}) {
+	n := len(pq.items)
+	item := x.(scheduledSync)
+	pq.index[item.changesetID] = n
+	pq.items = append(pq.items, item)
+}
+
+// Pop is not to be used directly, use heap.Pop(pq)
+func (pq *changesetPriorityQueue) Pop() interface{} {
+	item := pq.items[len(pq.items)-1]
+	delete(pq.index, item.changesetID)
+	pq.items = pq.items[:len(pq.items)-1]
+	return item
+}
+
+// Peek fetches the highest priority item without removing it
+func (pq *changesetPriorityQueue) Peek() (scheduledSync, bool) {
+	if len(pq.items) == 0 {
+		return scheduledSync{}, false
 	}
+	return pq.items[0], true
+}
 
-	ctx := context.Background()
-	ctx, q.cancel = context.WithCancel(ctx)
-	go func() {
-		for i := range schedule {
-			// Get most urgent changeset and sleep until it should be synced
-			now := time.Now()
-			nextSync := schedule[i].nextSync
-			d := nextSync.Sub(now)
-			timer := time.NewTimer(d)
-			select {
-			case <-ctx.Done():
-				timer.Stop()
-				return
-			case <-timer.C:
-				// Timer ready, try and send sync instruction
-			}
+// Upsert modifies at item if it exists or adds a new item
+// NOTE: If an existing item is high priority, it will not be changed back
+// to normal. This allows high priority items to stay that way through reschedules
+func (pq *changesetPriorityQueue) Upsert(s scheduledSync) {
+	i, ok := pq.index[s.changesetID]
+	if !ok {
+		heap.Push(pq, s)
+		return
+	}
+	oldPriority := pq.items[i].priority
+	pq.items[i] = s
+	if oldPriority == priorityHigh {
+		pq.items[i].priority = priorityHigh
+	}
+	heap.Fix(pq, i)
+}
 
-			select {
-			case <-ctx.Done():
-				return
-			case q.scheduled <- schedule[i].changesetID:
-			}
-		}
-	}()
+func (pq *changesetPriorityQueue) Get(id int64) (scheduledSync, bool) {
+	i, ok := pq.index[id]
+	if !ok {
+		return scheduledSync{}, false
+	}
+	item := pq.items[i]
+	return item, true
 }
 
-func (q *changesetQueue) enqueuePriority(ids []int64) error {
-	// q.priority will be buffered so that we allow
-	// a fixed depth in the high priority queue and
-	// can perform a non blocking add here
-	for _, id := range ids {
-		select {
-		case q.priority <- id:
-		default:
-			return errors.New("queue capacity for priority syncing reached")
-		}
+func (pq *changesetPriorityQueue) Reschedule(ss []scheduledSync) {
+	for i := range ss {
+		pq.Upsert(ss[i])
 	}
-	return nil
 }
 
-type syncSchedule struct {
-	changesetID int64
-	nextSync    time.Time
+func (pq *changesetPriorityQueue) Remove(id int64) {
+	i, ok := pq.index[id]
+	if ok {
+		heap.Remove(pq, i)
+	}
```

Comment:
nit: To be defensive and less cognitive load for the next person, would be nice to this
```suggestion
	i, ok := pq.index[id]
	if !ok {
		return
	}
	
	heap.Remove(pq, i)
```